### PR TITLE
Add font-family-no-missing-generic-family-keyword

### PR DESCRIPTION
**Pull request by stoyan**
This addresses issue (feature request) #2929. I mostly copied `font-family-no-duplicate-names` and tweaked accordingly.

To test I ran:

  $ npm test lib/rules/font-family-generic-fallback/__tests__/

Not sure how to improve the coverage of the tests. Results:

Test Suites: 1 passed, 1 total
Tests:       22 passed, 22 total
Snapshots:   0 total
Time:        2.609s
Ran all test suites matching /lib\/rules\/font-family-generic-fallback\/__tests__\//i.

=============================== Coverage summary ===============================
Statements   : 36.9% ( 2955/8008 )
Branches     : 4.85% ( 184/3794 )
Functions    : 5.15% ( 72/1398 )
Lines        : 37.12% ( 2950/7947 )
================================================================================
Jest: Coverage for statements (36.9%) does not meet global threshold (75%)
Jest: Coverage for branches (4.85%) does not meet global threshold (75%)
Jest: Coverage for lines (37.12%) does not meet global threshold (75%)
Jest: Coverage for functions (5.15%) does not meet global threshold (75%)
npm ERR! Test failed.  See above for more details.

> Which issue, if any, is this issue related to?

#2929

> Is there anything in the PR that needs further explanation?

I don't think so

### CODE REVIEW

**Comment by CAYdenberg** [comment]
Hey thanks for your idea and your work.

It's often a good idea to leave your issue up for a bit and see if there is any discussion ... the core is quite lively with hashing out issues and often things will be spotted that can avoid a chunk of work.

Having said that your tests do pass and your code runs fine: I don't think our coverage suite is really set up to report on one file at a time? If run `npm test` everything looks good (takes about 96 sec) ... while working you can use `npm run watch`, hit the `p` key, and then zero in on your test file using a regular expression.

**Comment by stoyan** [comment]
Thanks! Yeah, I poked around the code, and *then* read up on the procedures for pull request. My bad :)

**Comment by CAYdenberg** [comment]
NP. I like the idea and think this will get merged eventually.

File: lib/rules/font-family-generic-fallback/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# font-family-generic-fallback
+
+Require a generic fallback to any font-family declaration.
+
+```css
+a { font-family: Arial, sans-serif; }
+/**                        ↑
```
Comment (jeddy3) [inline lib/rules/font-family-generic-fallback/README.md]: Align the arrow to the start of the generic family

File: lib/rules/font-family-generic-fallback/README.md [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+```
+
+```css
+a { font: 1em/1.3 Arial; }
+```
+
+The following patterns are *not* considered violations:
+
+```css
+a { font-family: Times, serif; }
```
Comment (jeddy3) [inline lib/rules/font-family-generic-fallback/README.md]: Let's match these first two examples to the violations above, but fixed. I think that'll make it clearer.

File: lib/rules/font-family-generic-fallback/index.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      if (fontFamilies.length === 0) {
+        return;
+      }
+
+      const lastFamilyNode = fontFamilies[fontFamilies.length - 1];
+      if (isFamilyNameKeyword(lastFamilyNode)) {
+        return;
+      }
+
+      complain(
```
Comment (jeddy3) [inline lib/rules/font-family-generic-fallback/index.js]: No need for separate out the report into a `complain` function when it is only called once.

File: lib/rules/font-family-generic-fallback/index.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+const ruleName = "font-family-generic-fallback";
+
+const messages = ruleMessages(ruleName, {
+  generic: () => "Missing generic font family, e.g. serif"
+});
+
+const isFamilyNameKeyword = node =>
+  !node.quote && keywordSets.fontFamilyKeywords.has(node.value.toLowerCase());
+
```
Comment (CAYdenberg) [inline lib/rules/font-family-generic-fallback/index.js]: Does `!node.quote` mean anything in quotes will NOT be seen as a generic family name? I would have assumed that `font-family: Times, "serif"`, is perfectly valid and should pass, although it is slightly weird.

File: lib/rules/font-family-generic-fallback/README.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# font-family-generic-fallback
+
```
Comment (CAYdenberg) [inline lib/rules/font-family-generic-fallback/README.md]: As described in the [associated issue](https://github.com/stylelint/stylelint/issues/2929), please change all references to the rule name to font-family-no-missing-generic.

File: lib/rules/font-family-generic-fallback/index.js (hunk 4 above)
Comment (jeddy3) [reply lib/rules/font-family-generic-fallback/index.js]: According to [the spec](https://drafts.csswg.org/css-fonts/#font-family-prop), generic font families must be unquoted:

> The following generic family keywords are defined: ‘serif’, ‘sans-serif’, ‘cursive’, ‘fantasy’, and ‘monospace’. These keywords can be used as a general fallback mechanism when an author's desired font choices are not available. As keywords, they must not be quoted. Authors are encouraged to append a generic font family as a last alternative for improved robustness.

So, the code as it is LGTM.

Comment (CAYdenberg) [reply lib/rules/font-family-generic-fallback/index.js]: I stand corrected ... agreed.

File: lib/rules/font-family-generic-fallback/README.md (hunk 5 above)
Comment (jeddy3) [reply lib/rules/font-family-generic-fallback/README.md]: SGTM. As there seems to be [no objections](https://github.com/stylelint/stylelint/issues/2929#issuecomment-333479439) to the name proposal, @stoyan can you rename this `font-family-no-missing-generic`.

File: lib/rules/font-family-generic-fallback/README.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+Require a generic fallback to any font-family declaration.
```
Comment (jeddy3) [inline lib/rules/font-family-generic-fallback/README.md]: Can we change this to the following, as this is [now a `no` rules](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#single-line-descriptions):

Disallow missing generic families in lists of font family names.

File: lib/rules/font-family-generic-fallback/index.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+const findFontFamily = require("../../utils/findFontFamily");
+const keywordSets = require("../../reference/keywordSets");
+const report = require("../../utils/report");
+const ruleMessages = require("../../utils/ruleMessages");
+const validateOptions = require("../../utils/validateOptions");
+
+const ruleName = "font-family-generic-fallback";
+
+const messages = ruleMessages(ruleName, {
+  generic: () => "Missing generic font family, e.g. serif"
```
Comment (jeddy3) [inline lib/rules/font-family-generic-fallback/index.js]: For `no` rules, [the convention](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#determine-violation-messages) is to use `rejected`. A simple string would suffice too. Can you please use:

```
rejected: "Unexpected missing generic font family"
```

You'll need to update your tests accordingly too.

File: lib/rules/index.js [hunk 8]
```diff
@@ -52,6 +52,7 @@ const declarationPropertyUnitBlacklist = require("./declaration-property-unit-bl
 const declarationPropertyUnitWhitelist = require("./declaration-property-unit-whitelist");
 const declarationPropertyValueBlacklist = require("./declaration-property-value-blacklist");
 const declarationPropertyValueWhitelist = require("./declaration-property-value-whitelist");
+const fontFamilyGenericFallback = require("./font-family-generic-fallback");
```
Comment (jeddy3) [inline lib/rules/index.js]: Please update the rule name.

**Comment by stoyan** [comment]
Thanks folks, this is awesome! I'll work on it as soon as I can.

**Comment by jeddy3** [comment]
Changelog:

> -   Added: `font-family-no-missing-generic-family-keyword` rule ([#2930](https://github.com/stylelint/stylelint/pull/2930)).

**Pull request merged**